plugins/modules/ipacert.py

```diff
@@ -0,0 +1,366 @@
+        exit_args['certificate'] = ssl.DER_cert_to_PEM_cert(base64.b64decode(result['result']['certificate']))
+
+        # The root CA is last in the chain
+        exit_args['chain'] = [ssl.DER_cert_to_PEM_cert(c) for c in result['result']['certificate_chain'][:-1]]
+
+        exit_args['san_dnsname'] = [str(dnsname) for dnsname in result['result'].get('san_dnsname', [])]
+
+        for k in ['issuer', 'subject', 'serial_number']:
+            exit_args[k] = result['result'][k]
+
+        for k in ['valid_not_after', 'valid_not_before']:
+            exit_args[k] = parse_cert_timestamp(result['result'][k])
+
+    commands = [[csr, "cert_request", args]]
+    changed = ansible_module.execute_ipa_commands(commands, result_handler=result_handler, exception_handler=exception_handler)
+
+    return changed, exit_args
+
+
+def do_cert_revoke(ansible_module, ca, serial_number, reason):
+    args = {
+        'revocation_reason': reason,
+    }
+
+    if ca:
+        args['cacn'] = ca
+
+    commands = [[serial_number, "cert_revoke", args]]
+    changed = ansible_module.execute_ipa_commands(commands)
+
+    return changed, {}
+
+
+def do_cert_release(ansible_module, ca, serial_number):
+    args = {}
+
+    if ca:
+        args['cacn'] = ca
+
+    commands = [[serial_number, "cert_remove_hold", args]]
+    changed = ansible_module.execute_ipa_commands(commands)
+
+    return changed, {}
+
+
+def main():
+    ansible_module = IPAAnsibleModule(
+        argument_spec=dict(
+            # requested
+            csr=dict(type="str"),
+            principal=dict(type="str"),
+            add_principal=dict(type="bool"),
+            profile=dict(type="str"),
+
+            # revoked
+            reason=dict(type="str", choices=REVOCATION_REASONS.keys()),
+
+            # general
+            serial_number=dict(type="int"),
+            ca=dict(type="str"),
+
+            # state
+            state=dict(type="str", required=True,
+                       choices=["requested", "held", "released", "revoked"]),
+        ),
+        required_if=[
+            ('state', 'requested', ['csr', 'principal']),
+            ('state', 'held', ['serial_number']),
+            ('state', 'released', ['serial_number']),
+            ('state', 'revoked', ['serial_number', 'reason']),
+        ],
+        supports_check_mode=False,
+    )
+
+    ansible_module._ansible_debug = True
+
+    # Get parameters
+
+    # requested
+    csr = ansible_module.params_get("csr")
+    principal = ansible_module.params_get("principal")
+    add_principal = ansible_module.params_get("add_principal")
+    profile = ansible_module.params_get("profile")
+
+    # revoked
+    reason = ansible_module.params_get("reason")
+
+    # general
+    serial_number = ansible_module.params.get("serial_number")
+    ca = ansible_module.params_get("ca")  # XXX no test coverage
+
+    # state
+    state = ansible_module.params_get("state")
+
+    # Check parameters
+
+    invalid = []
+
```

Comment:
All ansible-freeipa modules should support the ipa_context variable, when `server` context is not supported, we use:

```python
if ansible_modulemodule.params_get("ipaapi_context") == "server":
    ansible_module.fail_json(
        msg="Context 'server' for ipacert is not yet supported."
    )
```

This will require `ipa_context` variable to be in the module README. Check module `ipavault` which also does not support `server` context.

Reply:
Thanks. I'll fix up the README in a separate pass, I need to write it from scratch now the module params etc are final-ish.